Python/ceval.c

```diff
@@ -3610,6 +3866,215 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
+                if (!interp->seq_abc) {
+                    goto error;
+                }
+            }
+            int match = PyObject_IsInstance(subject, interp->seq_abc);
+            if (match < 0) {
+                goto error;
+            }
+            PUSH(PyBool_FromLong(match));
+            DISPATCH();
+        }
+
+        case TARGET(MATCH_KEYS): {
+            // Final stage of mapping pattern matches (after Mapping and length
+            // checks): extract the desired values for further destructuring.
+            PyObject *keys = TOP();
+            PyObject *subject = SECOND();
+            assert(PyTuple_CheckExact(keys));
+            PyObject *values = match_keys(tstate, subject, keys);
+            if (!values) {
+                if (_PyErr_Occurred(tstate)) {
+                    goto error;
+                }
+                Py_INCREF(Py_False);
+                PUSH(Py_False);
+                DISPATCH();
+            }
+            // Success!
+            if (oparg) {
+                // If oparg is nonzero, collect the remaining items into a dict
+                // and put it on the stack where the subject used to be:
+                PyObject *rest = dict_without_keys(tstate, subject, keys);
+                if (!rest) {
+                    goto error;
+                }
+                assert(PyDict_CheckExact(rest));
+                SET_SECOND(rest);
+                Py_DECREF(subject);
+            }
+            SET_TOP(values);
+            Py_DECREF(keys);
+            Py_INCREF(Py_True);
+            PUSH(Py_True);
+            DISPATCH();
+        }
+
+        case TARGET(GET_INDEX): {
+            // PUSH(TOS[oparg])
+            PyObject *item = PySequence_GetItem(TOP(), oparg);
+            if (!item) {
+                goto error;
+            }
+            PUSH(item);
+            DISPATCH();
+        }
+
+        case TARGET(GET_INDEX_END): {
+            // PUSH(TOS[TOS1 - 1 - oparg])
+            // NOTE: We can't rely on support for negative indexing!
+            // Although PySequence_GetItem tries to correct negative indexes, we
+            // just use the length we already have at TOS1. In addition to
+            // avoiding tons of redundant __len__ calls, this also handles
+            // length changes during extraction more intuitively.
+            assert(PyLong_CheckExact(SECOND()));
+            Py_ssize_t len = PyLong_AsSsize_t(SECOND());
+            if (len < 0) {
+                goto error;
+            }
+            assert(0 <= len - 1 - oparg);
+            PyObject *item = PySequence_GetItem(TOP(), len - 1 - oparg);
+            if (!item) {
+                goto error;
+            }
+            PUSH(item);
+            DISPATCH();
+        }
+
+        case TARGET(GET_INDEX_SLICE): {
+            // PUSH(list(TOS[oparg & 0xFF: TOS1 - (oparg >> 8)]))
+            // NOTE: We can't rely on support for slicing or negative indexing!
```

Comment:
Why can't you use slicing, once you've done the bounds check?

Reply:
I don't think slicing (like negative indexing) is actually part of the `collections.abc.Sequence` protocol. `deque` doesn't support it, for example:

```py
>>> import collections
>>> dq = collections.deque(range(5))
>>> isinstance(dq, collections.abc.Sequence)
True
>>> dq[1:4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: sequence index must be integer, not 'slice'
```

I think it's reasonable to expect that a lot of Sequences out in the wild always expect `__getitem__` to pass an integer index. I still want those classes to work here.

Reply:
OK. That makes sense.
What about `UNPACK_EX`? That seems perfect for a sequence pattern with a star sub-pattern.

Reply:
> What about `UNPACK_EX`? That seems perfect for a sequence pattern with a star sub-pattern.

Yeah, I agree!

So (brainstorming here), we'll have two paths for compiling sequence patterns:
- One that uses `UNPACK_EX`, for cases where we have a non-wildcard star pattern.
- One that uses `BINARY_SUBSCR`, for everything else.

That way, patterns like `[first, *_, last]` stay efficient.

Reply:
I'd be inclined to use `UNPACK_EX` for `[first, *_, last]` as well.
I suspect that `[first, *_, last]` is a rare pattern, and the cost of the temporary list won't be that great.